lib/inspec/profile.rb

```diff
@@ -316,6 +316,25 @@ def check # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metri
       m_unsupported.each { |u| warn.call(meta_path, 0, 0, nil, "doesn't support: #{u}") }
       @logger.info 'Metadata OK.' if m_errors.empty? && m_unsupported.empty?
 
+      # verify that a lockfile is present if we have dependencies
+      if !metadata.dependencies.empty?
+        error.call(meta_path, 0, 0, nil, 'Profile need to be vendored with `inspec vendor`.') if !lockfile_exists?
+      end
+
+      if lockfile_exists?
+        # verify if metadata and lockfile are out of sync
+        if lockfile.deps.size != metadata.dependencies.size
+          error.call(meta_path, 0, 0, nil, 'inspec.yml and inspec.lock are out-of-sync. Please re-vendor with `inspec vendor`.')
+        end
+
+        metadata.dependencies.each { |dep|
+          # TODO: should we also verify that the soure is the same?
```

Comment:
And what about version constraints?

Reply:
Indeed. I left that open since it is more complicated. We should cover it, but I am not able to cover it as part of this PR.